### DISCUSSION TITLE
New token (after performing token refresh) not persisted when using Credentials provider

### DISCUSSION THREAD

**Discussion by stefanboz** [nextauthjs/next-auth#2003] [Help]
Hi there!

## My question

After making an API call to my back-end to get a new token, how can I tell NextAuth to continue using this new token instead of the original (old) one that is not valid anymore?

## What I'm trying to do

I'm using the latest version of NextAuth (3.21.1, but this happened with an older version too) and the Credentials provider. The API in my app works in a way that almost all API routes require a valid JWT token to be sent along with an API request. **I have to check if the JWT token is older than 12 hours and less than 24 hours old, and if so, make an API call to a route on my back-end with that token in order to get a new token.** 
The first time that getSession() is invoked, everything works fine and I get the new token. The issue is that that the new token is lost and the old token somehow appears whenever either getSession() or useSession() are invoked again in the app. That is to say, NextAuth continues using the original token that is not valid any more. As a consequence, I get an error when trying to refresh an invalid token and the user is logged out.

This is my **[...next-auth].js**:
```js

import NextAuth from 'next-auth';
import Providers from 'next-auth/providers';
import axios from 'axios';
import jwt_decode from 'jwt-decode';

const baseURL = process.env.NEXT_PUBLIC_API_URL;

async function refreshAccessToken(token) {
  try {
    const response = await axios({
      method: 'post',
      url: `${baseURL}/v1/auth/refresh`,
      headers: {
        'Content-Type': 'application/vnd.api+json',
        Accept: 'application/vnd.api+json',
        Authorization: `Bearer ${token}`,
      },
    });

    const refreshedToken = await response.data;

    if (response.status !== 200) {
      throw new Error('Token refresh failed: ', refreshedToken);
    }

    return {
      accessToken: refreshedToken.data['access_token'],
      expiresIn: refreshedToken.data['expires_in'],
    };
  } catch (error) {
    return {
      error: 'RefreshAccessTokenError',
    };
  }
}

const shouldRefreshToken = token => {
  const expirationDate = jwt_decode(token).exp * 1000;
  const currentTime = Date.now();

  return (
    expirationDate - currentTime <= 43200000 &&
    expirationDate - currentTime >= 0
  );
};

export default NextAuth({
  providers: [
    Providers.Credentials({
      name: 'Credentials',
      async authorize(credentials) {
        try {
          const response = await axios({
            url: `${baseURL}/v1/auth/login`,
            method: 'post',
            data: {
              email: credentials.email,
              password: credentials.password,
            },
            headers: {
              'Content-Type': 'application/vnd.api+json',
              Accept: 'application/vnd.api+json',
            },
          });

          const token = response.data.data['access_token'];

          return {
            accessToken: token,
          };
          
        } catch (error) {
          throw new Error(error);
        }
      },
    }),
  ],
  jwt: {
    secret: process.env.JWT_SECRET,
    signingKey: process.env.JWT_SIGNING_PRIVATE_KEY,
  },
  session: {
    jwt: true,
    maxAge: 24 * 60 * 60,
  },
  callbacks: {
    async jwt(token, user) {
      // the second time that either useSession() / getSession() are invoked the token here is the old (invalid) one
      if (account && user) {
        const expirationDate =
          jwt_decode(
            user.accessToken
          ).exp * 1000;

        token.accessToken =
          user.accessToken;
        token.accessTokenExpires = expirationDate;

        return {
          accessToken: user.accessToken,
          accessTokenExpires: expirationDate,
        };
      }
      
      if (Date.now() < token.accessTokenExpires) {
        const shouldRefresh = shouldRefreshToken(token.accessToken);

        if (shouldRefresh) {
            // this is the new token that NextAuth should continue using instead of the old one
            const newToken = await refreshAccessToken(token.accessToken);

            const decodedToken = jwt_decode(newToken.accessToken);

            token.accessToken = newToken.accessToken;
            token.accessTokenExpires = decodedToken.exp * 1000;
            token.exp = decodedToken.exp;
            token.iat = decodedToken.iat;

            return Promise.resolve(token);
          
        } else {
          // token is valid
          return Promise.resolve(token);
        }
      } else {
        // token has expired, log the user out
        token = null;
        return Promise.resolve(token);
      }
    },
    async session(session, token) {
      if (token) {
        session.accessToken = token.accessToken;
        session.accessTokenExpires = token.accessTokenExpires;
      }

      return Promise.resolve(session);
    },
  },
});

```

And in my **_app.js** I use NextAuth's Provider to wrap my whole app with it, like this:
```js

<Provider session={pageProps.session}>
   // other code
</Provider>

```

And this is the **getServerSideProps()** on my app's landing page, the **index.js**:
```js

export async function getServerSideProps(context) {
  try {
    const session = await getSession(context);

    if (session) {
      // the user was logged in previously, I check token validity here and if everything is OK I redirect the user to the /home route
      return {
        redirect: {
          destination: '/home',
          permanent: false,
        },
      };
    }

    const response = await http({
      method: 'get',
      url: `/v1/profiles?page[number]=1`,
    });
    return {
      props: {
        data: response.data,
      },
    };
  } catch (error) {
    return {
      props: {
        data: null,
      },
    };
  }
}

```

I tried pretty much anything I could think of, I read the docs and the existing issues here on GitHub, but to no avail. Can anyone please tell if I'm doing something wrong?

Thanks in advance and thanks for reading this!

**Comment by chomii** [comment]
Hi. I have exactly the same issue. I don't have the solution but I hope I will find it here.

**Comment by makezi** [comment]
It is the same issue we're having here: #1455 but there doesn't seem to be a solution just yet.

**Reply by stefanboz** [reply]
Thanks @makezi 

Wow, so it is a huge bug after all. I don't want to be logging my users out all the time, it's practically rendering NextAuth useless for me

**Reply by balazsorban44** [reply]
Wouldn't call it a bug, just a not yet implemented feature. 🙂 The tutorial and the solutions that are circulating around are based off of my suggestions from a while back, and it seems that I haven't been able to think of something that causes people problems. I'm open for suggestions on how to improve it, but currently I'm very busy with other features and general maintenance of the library. I come back to it from time to time and try to think of how it could be fixed.

**Reply by ajaykarthikr** [reply]
Hey! Even I am facing the exact same issue. Is there any timeline when this issue might be fixed? I saw there's a PR at #2072  which could potentially resolve this issue.

**Comment by HarunKilic** [comment]
Did you find any workaround or fix regarding to this? I am having exact same issue.

**Reply by stefanboz** [reply]
Nope, I ditched NextAuth. Too big of an issue for me

**Reply by HarunKilic** [reply]
What solution did u go with ?

**Reply by stefanboz** [reply]
Custom authentication. I have a backend though, and when I log user in I keep the session data in a cookie and that's basically that.

**Reply by HarunKilic** [reply]
Did also try to use custom authentication. But handling serverside and clientside is a headache. Specially when you have needed informations.

**Comment by mikenieva** [comment]
Living the same problem.

**Comment by mackankowski** [comment]
I think we've experienced the same issue: 

https://github.com/nextauthjs/next-auth/discussions/2850

I'm afraid that returning to CRA it's our last hope.

**Reply by balazsorban44** [reply]
https://github.com/nextauthjs/next-auth/discussions/2850#discussioncomment-1395106
